src/components/ClassSearch.vue

```diff
@@ -133,84 +282,33 @@ export default {
-        'communication_requirement': getRegexFuncs(this.chosenFilters.ciInput),
-        'level': getRegexFuncs(this.chosenFilters.levelInput),
-        'total_units': getRegexFuncs(this.chosenFilters.unitInput)
-      };
-        // and or or function based on filter mode
-      const filterAction = this.filterGroupModes[this.filterGroupMode];
+      // Filter subjects that match all filter sets and the text filter
       const filteredSubjects = this.allSubjects.filter(function (subject) {
-        for (const attrs in filters) {
-          // each test function in a filter group
-          const testers = filters[attrs];
-          if (testers.length) {
-            // if a single attribute group in a set returns true, the filter will match it
-            let passesAnyAttributeGroupInSet = false;
-            const attrSet = attrs.split(',');
-            for (let a = 0; a < attrSet.length; a++) {
-              const attr = attrSet[a];
-              let subjectattr = subject[attr];
-              if (!subject[attr]) {
-                subjectattr = '';
-              }
-              // start with false for OR mode, and true for AND mode
-              let passesAttributeGroup = !filterAction(false, true);
-              // use the filter mode function (OR or AND) and test all filters in a group
-              for (let t = 0; t < testers.length; t++) {
-                passesAttributeGroup = filterAction(passesAttributeGroup, testers[t](subjectattr));
-              }
-              if (passesAttributeGroup) {
-                passesAnyAttributeGroupInSet = true;
-              }
-            }
-            // if the subject passes no attribute group in the set, don't include it
-            if (!passesAnyAttributeGroupInSet) {
-              return false;
-            }
-          }
+        var matches = true;
+        for (var filterGroup in this.allFilters) {
+          matches = matches && this.allFilters[filterGroup].matches(subject, this.chosenFilters[filterGroup], { nameInput: this.nameInput });
         }
-        return true;
-      });
-      if (this.chosenFilters.nameInput.length) {
-        const sortingOrder = [this.chosenFilters.nameInput, '^' + this.chosenFilters.nameInput, escapeRegExp(this.chosenFilters.nameInput), '^' + escapeRegExp(this.chosenFilters.nameInput)];
-        const sortingFuncs = getRegexFuncs(sortingOrder);
-        const getOrderForString = function (matchingString) {
-          const matches = sortingFuncs.map((func) => func(matchingString));
-          return matches.lastIndexOf(true);
-        };
-        const getOrder = function (subject) {
-          const idMatch = getOrderForString(subject.subject_id);
-          const nameMatch = getOrderForString(subject.title);
-          return Math.max(idMatch, nameMatch);
-        };
+        matches = matches & textFilter.matches(subject, { nameInput: this.nameInput });
+        return matches;
+      }.bind(this));
```

Comment:
This looks like a natural `all(Object.keys(this.allFilters).map(...))` kind of thing, if you feel like it. This way is fine too though if you think it's better.